### PR TITLE
Refactor linear scan

### PR DESCRIPTION
**Pull request by qinsoon**
This PR closes https://github.com/mmtk/mmtk-core/issues/495 and closes https://github.com/mmtk/mmtk-core/issues/485.

This PR:
* adds `ObjectIterator` that helps iterate though objects in an address range using alloc bit, and a related trait `LinearScanObjectSize`.
* uses `ObjectIteartor` in mark sweep and mark compact
* removes current linear scan code (which is unused)
* refactors chunk sweep code for mark sweep to make it clearer (https://github.com/mmtk/mmtk-core/issues/485)

This PR has no observable performance difference for marksweep. The graph below shows the time for sweeping chunk.
WX20220113-133046

### CODE REVIEW

**Comment by k-sareen** [comment]
Quick question -- did you benchmark the PR with both in header and on side mark bits?

**Comment by qinsoon** [comment]
> Quick question -- did you benchmark the PR with both in header and on side mark bits?

No. The graph only shows the performance for in header mark bits. I can measure the other if you think it might be different.

**Comment by k-sareen** [comment]
The default configuration is the bulk load from memory, right? Maybe you can run a very small (~10 invocations) benchmark result for that if possible

**Comment by qinsoon** [comment]
> The default configuration is the bulk load from memory, right? Maybe you can run a very small (~10 invocations) benchmark result for that if possible

Yes, the default is bulk load with on side mark bits. That sounds good. I will do another run and probably will get the results back during the weekend.

File: src/util/linear_scan.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+use crate::util::ObjectReference;
+use crate::vm::ObjectModel;
+use crate::vm::VMBinding;
+use std::marker::PhantomData;
+
+/// Iterate over an address range, and find each object by alloc bit.
+/// ATOMIC_LOAD_ALLOC_BIT can be set to false if it is known that loading alloc bit
+/// non-atomically is correct (e.g. a single thread is scanning this address range, and
+/// it is the only thread that accesses alloc bit).
+pub struct ObjectIterator<VM: VMBinding, S: LinearScanObjectSize, const ATOMIC_LOAD_ALLOC_BIT: bool>
```
Comment (k-sareen) [inline src/util/linear_scan.rs:12]: I was wondering if the type parameter `LinearScanObjectSize` is required? Would a function pointer to a `size()` function not be good enough? Or are you trying to reduce overhead of indirect function calls? (Is it even significant enough?). Either way, it's not a big deal, but was wondering about your thoughts.

**Review by k-sareen — COMMENTED** [review]
LGTM other than the one question (will approve after discussion and results of bulk load mark bit). Nice work :)

File: src/util/linear_scan.rs (hunk 1 above)
Comment (qinsoon) [reply src/util/linear_scan.rs:12]: Yeah, I was considering about performance and trying to be safe, as we know using trait would have no overhead at all. Maybe using a function pointer is enough.

**Comment by qinsoon** [comment]
@k-sareen The results are here.

For mark bits in the header,
ms-header

For mark bits on the side,
ms-side

**Review by k-sareen — APPROVED** [review]

**Pull request merged**